### PR TITLE
Added cheats to the chat commands

### PR DESCRIPTION
**Pull request by gggfreak2003**
Added cheats to RTTR.

The cheats has to be activated by entering "winter". After the cheats are activated, they could be entered as chat commands (Press ENTER and then "!gold" for 10 units gold). Valid commands are: ironore, beer, tongs, hammer, axe, saw, pickaxe, shovel, crucible, rodandline, scythe, cleaver, rollingpin, bow, boat, iron, flour, fish, bread, wood, boards, stones, grain, coins, gold, coal, meat, ham, water. The command "givemeall" gives you 10 units of all goods.

### CODE REVIEW

File: src/desktops/dskGameInterface.cpp [hunk 1]
```diff
@@ -689,6 +689,9 @@ bool dskGameInterface::Msg_KeyDown(const KeyEvent& ke)
                 {
                     isCheatModeOn = !isCheatModeOn;
                     curCheatTxt.clear();
+                    char text[256];
+                    snprintf(text, sizeof(text), "Cheat mode is on");
```
Comment (baka0815) [inline src/desktops/dskGameInterface.cpp]: Shouldn't this check for ``isCheatModeOn``?

File: src/desktops/dskGameInterface.cpp [hunk 2]
```diff
@@ -1016,7 +1019,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
             const_cast<GameWorld&>(game_->world).GetDescriptionWriteable() = newDesc;
             worldViewer.InitTerrainRenderer();
         }
-    }
+    } else if(cmd == "ironore")	cheatAdd(GD_IRONORE, 10, cmd);
```
Comment (baka0815) [inline src/desktops/dskGameInterface.cpp]: Line break before the ``else``?

File: src/desktops/dskGameInterface.cpp [hunk 3]
```diff
@@ -1016,7 +1019,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+
+		// Create a list and initialize it with vector
+		std::list<GoodType> listOfInts(arr, arr + sizeof(arr) / sizeof(int));
+
+		// Iterate over the list and display numbers
+		for (GoodType val : listOfInts)
+			cheatAdd(val, 10, "");
+
+		char text[256];
+		snprintf(text, sizeof(text), "Cheat: Added 10 of all goods");
```
Comment (baka0815) [inline src/desktops/dskGameInterface.cpp]: Should be translated (see ``cheatAdd``):
````cpp
snprintf(text, sizeof(text), _("Cheat: Added 10 of all goods"));
````

File: src/desktops/dskGameInterface.cpp (hunk 1 above)
Comment (gggfreak2003) [reply src/desktops/dskGameInterface.cpp]: You are right. Thanks

File: src/desktops/dskGameInterface.cpp (hunk 3 above)
Comment (gggfreak2003) [reply src/desktops/dskGameInterface.cpp]: How I could add a translation?

Comment (Flow86) [reply src/desktops/dskGameInterface.cpp]: ```_("Text to translate")```

File: src/desktops/dskGameInterface.cpp [hunk 4]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
             worldViewer.InitTerrainRenderer();
         }
     }
+    else if(cmd == "ironore")	cheatAdd(GD_IRONORE, 10, cmd);
```
Comment (Flow86) [inline src/desktops/dskGameInterface.cpp]: isnt there a way to the the name from an GD_... constant? so no every good type works then. would be much better and less error prone

File: src/desktops/dskGameInterface.cpp [hunk 5]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+
+		// Create a list and initialize it with vector
+		std::list<GoodType> listOfInts(arr, arr + sizeof(arr) / sizeof(int));
+
+		// Iterate over the list and display numbers
+		for (GoodType val : listOfInts)
+			cheatAdd(val, value, "");
+
+		char text[256];
+		snprintf(text, sizeof(text), _("Cheat: Added %d of all goods"), value);
```
Comment (Shawn8901) [inline src/desktops/dskGameInterface.cpp]: I am not sure which way is better, but all other places in the codebase use `%` to replace the placeholder.

File: src/desktops/dskGameInterface.cpp [hunk 6]
```diff
@@ -689,6 +689,9 @@ bool dskGameInterface::Msg_KeyDown(const KeyEvent& ke)
                 {
                     isCheatModeOn = !isCheatModeOn;
                     curCheatTxt.clear();
+                    char text[256];
+                    snprintf(text, sizeof(text), _("Cheat mode is %s"), isCheatModeOn ? _("on") : _("off"));
```
Comment (Shawn8901) [inline src/desktops/dskGameInterface.cpp]: i would here also prefer the `%`-operator instead the snprintf

Comment (Flamefire) [reply src/desktops/dskGameInterface.cpp]: Don't use snprintf at all. Use Boost format which is safer. BTW: No need for this message. There already is an icon showing whether cheat mode is on

File: src/desktops/dskGameInterface.cpp [hunk 7]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+    else if(cmd == "boards") cheatAdd(GD_BOARDS, 10, cmd);
+    else if(cmd == "stones") cheatAdd(GD_STONES, 10, cmd);
+    else if(cmd == "grain") cheatAdd(GD_GRAIN, 10, cmd);
+    else if(cmd == "coins") cheatAdd(GD_COINS, 10, cmd);
+    else if(cmd == "gold") cheatAdd(GD_GOLD, 10, cmd);
+    else if(cmd == "coal") cheatAdd(GD_COAL, 10, cmd);
+    else if(cmd == "meat") cheatAdd(GD_MEAT, 10, cmd);
+    else if(cmd == "ham") cheatAdd(GD_HAM, 10, cmd);
+    else if(cmd == "water") cheatAdd(GD_WATER, 10, cmd);
+    else if(cmd == "givemeall") cheatAddAll();
```
Comment (Flamefire) [inline src/desktops/dskGameInterface.cpp]: You could use the `WARE_NAMES` array to map a string to a good.

File: src/desktops/dskGameInterface.cpp [hunk 8]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+
+void dskGameInterface::cheatAdd(GoodType type, unsigned value, const std::string& type_string) {
+	if(isCheatModeOn) {
+		GameWorldBase& world = worldViewer.GetWorldNonConst();
+		GamePlayer& player = world.GetPlayer(worldViewer.GetPlayerId());
+
+		nobBaseWarehouse* wh = player.GetFirstWH();
+		Inventory goods;
+		goods.Add(type, value);
+		wh->AddGoods(goods, true);
```
Comment (Flamefire) [inline src/desktops/dskGameInterface.cpp]: This only works for singleplayer and will async in replays and multiplayer. Correct way would be to add a GameCommand for that, especially as the viewer should not change the game (`GetWorldNonConst` is a work-around for the current messy code base)

File: src/desktops/dskGameInterface.cpp [hunk 9]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+
+		if(type_string.length() != 0) {
+			char text[256];
+			snprintf(text, sizeof(text), _("Cheat: Added %d %s"), value, type_string.c_str());
```
Comment (Flamefire) [inline src/desktops/dskGameInterface.cpp]: Use `Boost.Format`

File: src/desktops/dskGameInterface.cpp [hunk 10]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+			messenger.AddMessage("", 0, CD_SYSTEM, text, COLOR_RED);
+		}
+	}
+}
+
+void dskGameInterface::cheatAddAll(unsigned value=10){
```
Comment (Flamefire) [inline src/desktops/dskGameInterface.cpp]: Default value in definition is not allowed (Compilation should fail)

File: src/desktops/dskGameInterface.cpp [hunk 11]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+				GD_PICKAXE, GD_SHOVEL, GD_CRUCIBLE, GD_RODANDLINE, GD_SCYTHE,
+				GD_CLEAVER, GD_ROLLINGPIN, GD_BOW, GD_BOAT, GD_IRON, GD_FLOUR,
+				GD_FISH, GD_BREAD, GD_WOOD, GD_BOARDS, GD_STONES, GD_GRAIN,
+				GD_COINS, GD_GOLD, GD_COAL, GD_MEAT, GD_HAM, GD_WATER };
+
+		// Create a list and initialize it with vector
+		std::list<GoodType> listOfInts(arr, arr + sizeof(arr) / sizeof(int));
+
+		// Iterate over the list and display numbers
+		for (GoodType val : listOfInts)
```
Comment (Flamefire) [inline src/desktops/dskGameInterface.cpp]: C++11 is not possible ATM, use BOOST_FOREACH

File: src/desktops/dskGameInterface.cpp [hunk 12]
```diff
@@ -1017,6 +1020,75 @@ void dskGameInterface::OnChatCommand(const std::string& cmd)
...
+			char text[256];
+			snprintf(text, sizeof(text), _("Cheat: Added %d %s"), value, type_string.c_str());
+			messenger.AddMessage("", 0, CD_SYSTEM, text, COLOR_RED);
+		}
+	}
+}
+
+void dskGameInterface::cheatAddAll(unsigned value=10){
+	if(isCheatModeOn) {
+		GoodType arr[] = { GD_IRONORE, GD_BEER, GD_TONGS, GD_HAMMER, GD_AXE, GD_SAW,
```
Comment (Flamefire) [inline src/desktops/dskGameInterface.cpp]: Use a `boost::array` with a fixed size. This also saves the copy to a list below.

**Comment by Flow86** [comment]
any progress here?

File: src/desktops/dskGameInterface.cpp (hunk 4 above)
Comment (Flamefire) [reply src/desktops/dskGameInterface.cpp]: see https://github.com/Return-To-The-Roots/s25client/pull/865#discussion_r214508949 `WARE_NAMES`

**Comment by Flow86** [comment]
stale for a long time

**Pull request closed without merging**